dmcontent/questions.py

```diff
@@ -480,6 +479,53 @@ def summary(self, service_data):
         return ListSummary(self, service_data)
 
 
+class Hierarchy(List):
+    """
+    For our purposes, a Hierarchy is like a List, except the entries
+    are potentially related to each other in a "subsumptive
+    containment hierarchy". For example, every service that is a member
+    of a child category is _by definition_ also a member of its parent
+    category as well, and we can add those parent categories as a
+    denormalization to make searching more efficient.
+    """
+
+    def _get_data(self, form_data):
+        if self.id not in form_data:
+            return {self.id: None}
+
+        values = set(form_data.getlist(self.id))
+        values.update(self.get_missing_values(values))
+
+        return {self.id: sorted(values) or None}
+
+    def summary(self, service_data):
+        return HierarchySummary(self, service_data)
+
+    def get_missing_values(self, selected_values_set):
+        """
+        Recursively retrieves un-selected parent categories of the
+        passed-in selection, as a set of 'value' strings.
+        :param selected_values_set: initially-selected categories (e.g. by the user)
+        :return: additional values that should also be selected
+        """
+        expected_values = set()
+
+        def set_expected_values(options, parents=None):
+            if parents is None:
+                parents = list()
+
+            for option in options:
+                value = option.get('value', option.get('label'))
+                if value in selected_values_set:
+                    expected_values.update(parents)
+                children = option.get('options')
+                if children:
+                    set_expected_values(children, parents + [value])
+
+        set_expected_values(self._data.get('options', []))
+        return expected_values - selected_values_set
```

Comment:
@galund I'd put it in the commit message related to this change.

This is a minor point, but I'd use an argument instead of relying on a closure. It makes clearer that this is a recursive function with an accumulator and will make it easier to move it out if we need to since it's self-contained.

So something like:
```python

def get_expected_values(expected_values, options, parents=None):
    ...
    if children:
        get_expected_values(expected_values, children, parents + [value])
    ...

return get_expected_values(set(), self._data.get('options', [])) - selected_values_set
```